### PR TITLE
Move interface expansion in hierarchy.cc into a helper class

### PR DESCRIPTION
**Pull request by rswarbrick**
There should be no functional change, but this splits up the control
flow across functions, using class fields to hold the state that's
being tracked. The result should be a bit easier to read.

This is part of work to add bind support, but I'm doing some
refactoring in the hierarchy pass to make the code a bit easier to
work with. The idea is that (eventually) the IFExpander object will
hold all the logic for expanding interfaces, and then other code can
do bind insertion.

(Note: This patch is part of the larger queue at #2752. The idea is to split things out to avoid having a giant PR to review all at once)

### CODE REVIEW

**Comment by rswarbrick** [comment]
Ping! Any chance someone could take a look at this code?

**Review by zachjs — APPROVED** [review]
I haven't compared line by line, but this indeed appears equivalent to the existing implementation. While it still leaves the handling of interfaces in rough state, I would argue it's an improvement. I will merge this tomorrow unless there are objections from other maintainers.

**Pull request merged**
